### PR TITLE
Conversational dataset support for `CPOTrainer`

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Part of #2071

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/cpo_trainer.mdx [hunk 1]
```diff
@@ -1,100 +1,65 @@
...
+tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
+train_dataset = load_dataset("trl-lib/Capybara-Preferences", split="train")
 
-cpo_trainer = CPOTrainer(
-    model,
-    args=training_args,
-    train_dataset=train_dataset,
-    tokenizer=tokenizer,
-)
+training_args = CPOConfig(output_dir="cpo-qwen2", logging_steps=10)
```
Comment (lewtun) [inline docs/source/cpo_trainer.mdx:32]: turbo nit:

```suggestion
training_args = CPOConfig(output_dir="Qwen2-0.5B-CPO", logging_steps=10)
```

File: docs/source/cpo_trainer.mdx [hunk 2]
```diff
@@ -1,100 +1,65 @@
...
+from trl import CPOConfig, CPOTrainer
+from transformers import AutoModelForCausalLM, AutoTokenizer
 
-```py
-training_args = CPOConfig(
-    beta=0.1,
-)
+model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
+tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
+train_dataset = load_dataset("trl-lib/Capybara-Preferences", split="train")
```
Comment (lewtun) [inline docs/source/cpo_trainer.mdx:30]: Not a strong opinion, but should we use `trl-lib/ultrafeedback_binarized` for all our preference datasets? It has the nice feature of being well tested in the community and usually "just works" when it comes to checking a preference method. Capyrbara-Preferences has the nice feature of being multi-step preference but less well known

File: docs/source/cpo_trainer.mdx [hunk 3]
```diff
@@ -104,6 +69,34 @@ While training and evaluating we record the following reward metrics:
 * `rewards/margins`: the mean difference between the chosen and corresponding rejected rewards
 * `nll_loss`: the mean negative log likelihood loss of the policy model for the chosen responses
 
+## CPO variants
```
Comment (lewtun) [inline docs/source/cpo_trainer.mdx:72]: Really nice docs!

File: docs/source/cpo_trainer.mdx [hunk 4]
```diff
@@ -104,6 +69,34 @@ While training and evaluating we record the following reward metrics:
...
+
+### SimPO
```
Comment (lewtun) [inline docs/source/cpo_trainer.mdx:74]:
```suggestion
### Simple Preference Optimization (SimPO)
```

File: tests/test_cpo_trainer.py [hunk 5]
```diff
@@ -37,15 +37,16 @@ def setUp(self):
 
     @parameterized.expand(
         [
-            ["gpt2", "sigmoid"],
-            ["t5", "hinge"],
-            ["gpt2", "ipo"],
-            ["t5", "ipo"],
-            ["gpt2", "simpo"],
-            ["t5", "simpo"],
+            ["gpt2", "sigmoid", "standard_preference"],
```
Comment (lewtun) [inline tests/test_cpo_trainer.py:40]: At some point we should change this to replace `gpt2` with small instruct variants like `Qwen2.5-0.5B-Instruct` and `Llama-3.2-1B-Instruct` (although the latter is gated which means we can skip it). Applying CPO and other DPO-like methods to base models doesn't make sense IMO

File: trl/trainer/cpo_config.py [hunk 6]
```diff
@@ -27,6 +27,9 @@ class CPOConfig(TrainingArguments):
     command line.
 
     Parameters:
+        learning_rate (`float`, *optional*, defaults to `1e-6`):
```
Comment (lewtun) [inline trl/trainer/cpo_config.py:30]: Nice!

**Review by lewtun — APPROVED** [review]
Thanks for the nice clean up and refactor! LGTM with some minor nits

File: docs/source/cpo_trainer.mdx (hunk 2 above)
Comment (qgallouedec) [reply docs/source/cpo_trainer.mdx:30]: #2176

File: tests/test_cpo_trainer.py (hunk 5 above)
Comment (qgallouedec) [reply tests/test_cpo_trainer.py:40]: #2177
